src/rust/engine/graph/src/entry.rs

```diff
@@ -435,9 +454,9 @@ impl<N: Node> Entry<N> {
         ..
       } => {
         if result == Some(Err(N::Error::invalidated())) {
```

Comment:
I believe that it should be possible to merge both the `invalidated` and `Errored` cases in this if-else, as neither is now persisted.

BUT: rather than "dropping" the `waiters` (from the `Running` struct), as in the first case, probably what you'll want to do is just always send the error value. The effect of dropping the `waiters` is that anyone waiting for the value of the node will get an invalidated error value, here: https://github.com/pantsbuild/pants/blob/17e2401e4e0a0cec602f88f1b34be9fb1b8472a6/src/rust/engine/graph/src/entry.rs#L303-L306 ... actually _sending_ them the invalidated Err value (which is what you would do if you didn't do this `result == Some(Err(N::Error::invalidated()))`) would mean that the future would be satisfied successfully with the invalidated value (ie, it would have the same effect as dropping the waiter). Not sending the Errored value makes the patch incorrectly currently.

So: all this to say: you should be able to merge the invalidated/Errored blocks as long as you actually _send_ the error value that you got (the same as how we send a successful value here: https://github.com/pantsbuild/pants/blob/17e2401e4e0a0cec602f88f1b34be9fb1b8472a6/src/rust/engine/graph/src/entry.rs#L498-L500).

Reply:
So I've done that here, but it's not clear to me why this works without the `waiter.send()` part added. Is there a test I can write to validate the behavior you describe?